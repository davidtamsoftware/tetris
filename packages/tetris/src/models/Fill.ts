export enum Fill {
    Blank,
    Orange,
    Yellow,
    Green,
    Blue,
    Gray,
    Red,
    Purple,
    White,
}

export default Fill;
