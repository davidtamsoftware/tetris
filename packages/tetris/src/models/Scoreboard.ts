export interface Scoreboard {
    score: number;
    highscore: number;
    level: number;
    lines: number;
}
