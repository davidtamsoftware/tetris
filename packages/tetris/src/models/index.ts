export * from "./piece";
export * from "./playField";
