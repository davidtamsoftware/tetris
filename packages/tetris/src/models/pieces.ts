export type Piece = [
    [number, number, number, number],
    [number, number, number, number],
    [number, number, number, number],
    [number, number, number, number]
];

// 10W X 18H
export type GameState = [
    [number, number, number, number, number, number, number, number, number, number],
    [number, number, number, number, number, number, number, number, number, number],
    [number, number, number, number, number, number, number, number, number, number],
    [number, number, number, number, number, number, number, number, number, number],
    [number, number, number, number, number, number, number, number, number, number],
    [number, number, number, number, number, number, number, number, number, number],
    [number, number, number, number, number, number, number, number, number, number],
    [number, number, number, number, number, number, number, number, number, number],
    [number, number, number, number, number, number, number, number, number, number],
    [number, number, number, number, number, number, number, number, number, number],
    [number, number, number, number, number, number, number, number, number, number],
    [number, number, number, number, number, number, number, number, number, number],
    [number, number, number, number, number, number, number, number, number, number],
    [number, number, number, number, number, number, number, number, number, number],
    [number, number, number, number, number, number, number, number, number, number],
    [number, number, number, number, number, number, number, number, number, number],
    [number, number, number, number, number, number, number, number, number, number],
    [number, number, number, number, number, number, number, number, number, number]
];

export const o: Piece = [
    [0,0,0,0],
    [0,1,1,0],
    [0,1,1,0],
    [0,0,0,0],
];

export const i: Piece = [
    [0,0,0,0],
    [2,2,2,2],
    [0,0,0,0],
    [0,0,0,0],
];

export const s: Piece = [
    [0,0,0,0],
    [0,0,3,3],
    [0,3,3,0],
    [0,0,0,0],
];

export const z: Piece = [
    [0,0,0,0],
    [0,4,4,0],
    [0,0,4,4],
    [0,0,0,0],
];

export const l: Piece = [
    [0,0,0,0],
    [0,5,5,5],
    [0,5,0,0],
    [0,0,0,0],
];

export const j: Piece = [
    [0,0,0,0],
    [0,6,6,6],
    [0,0,0,6],
    [0,0,0,0],
];

export const t: Piece = [
    [0,0,0,0],
    [0,7,7,7],
    [0,0,7,0],
    [0,0,0,0],
];

export const pieces: Piece[] = [o, i, s, z, l, j, t];