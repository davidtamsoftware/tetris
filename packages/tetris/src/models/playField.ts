export type PlayField = number[][];

export const generatePlayField = (row: number, col: number) => new Array<number[]>(row).fill(new Array<number>(col).fill(0));

export const playField: PlayField = generatePlayField(18, 10);
