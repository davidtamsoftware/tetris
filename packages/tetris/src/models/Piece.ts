import { Fill } from "./Fill";

export type Piece = Fill[][];

export const o: Piece = [
    [Fill.Blue, Fill.Blue],
    [Fill.Blue, Fill.Blue],
];

export const i: Piece = [
    [Fill.Blank, Fill.Blank, Fill.Blank, Fill.Blank],
    [Fill.Gray, Fill.Gray, Fill.Gray, Fill.Gray],
    [Fill.Blank, Fill.Blank, Fill.Blank, Fill.Blank],
    [Fill.Blank, Fill.Blank, Fill.Blank, Fill.Blank],
];

export const s: Piece = [
    [Fill.Blank, Fill.Blank, Fill.Blank],
    [Fill.Blank, Fill.Yellow, Fill.Yellow],
    [Fill.Yellow, Fill.Yellow, Fill.Blank],
];

export const z: Piece = [
    [Fill.Blank, Fill.Blank, Fill.Blank],
    [Fill.Orange, Fill.Orange, Fill.Blank],
    [Fill.Blank, Fill.Orange, Fill.Orange],
];

export const l: Piece = [
    [Fill.Blank, Fill.Blank, Fill.Blank],
    [Fill.Green, Fill.Green, Fill.Green],
    [Fill.Green, Fill.Blank, Fill.Blank],
];

export const j: Piece = [
    [Fill.Blank, Fill.Blank, Fill.Blank],
    [Fill.Red, Fill.Red, Fill.Red],
    [Fill.Blank, Fill.Blank, Fill.Red],
];

export const t: Piece = [
    [Fill.Blank, Fill.Blank, Fill.Blank],
    [Fill.Purple, Fill.Purple, Fill.Purple],
    [Fill.Blank, Fill.Purple, Fill.Blank],
];

export const pieces: Piece[] = [o, i, s, z, l, j, t];

export default Piece;
