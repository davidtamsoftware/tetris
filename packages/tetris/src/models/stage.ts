import * as pieces from "./pieces";
import { GameState as State, Piece } from "./pieces";

export interface PositionGrid {
    row: number,
    col: number,
}

export const rotateRight = (state: State, position: PositionGrid, p: Piece) => {
    // rotate
    // check for collision
    // return result if no collision
    // if collision move left 1
    // check for collision
    // return result if no collision
    // if collision move right 1
    // check for collision
    // return result if no collision
    // else return same result as args

    const piece: Piece = [
        [p[3][0], p[2][0], p[1][0], p[0][0]],
        [p[3][1], p[2][1], p[1][1], p[0][1]],
        [p[3][2], p[2][2], p[1][2], p[0][2]],
        [p[3][3], p[2][3], p[1][3], p[0][3]],
    ];

    if (!hasCollision(state, position, piece)) {
        return {
            state,
            position,
            piece,
        };
    }

    const newPosL = { row: position.row, col: position.col - 1 };
    if (!hasCollision(state, newPosL, piece)) {
        return {
            state,
            position: newPosL,
            piece,
        };
    }

    const newPosR = { row: position.row, col: position.col + 1 };
    if (!hasCollision(state, newPosR, piece)) {
        return {
            state,
            position: newPosR,
            piece,
        };
    }

    return {
        state,
        position,
        piece: p,
    }
};

export const rotateLeft = (state: State, position: PositionGrid, p: Piece) => {
    // rotate
    // check for collision
    // return result if no collision
    // if collision move left 1
    // check for collision
    // return result if no collision
    // if collision move right 1
    // check for collision
    // return result if no collision
    // else return same result as args

    const piece: Piece = [
        [p[0][3], p[1][3], p[2][3], p[3][3]],
        [p[0][2], p[1][2], p[2][2], p[3][2]],
        [p[0][1], p[1][1], p[2][1], p[3][1]],
        [p[0][0], p[1][0], p[2][0], p[3][0]],
    ]

    if (!hasCollision(state, position, piece)) {
        return {
            state,
            position,
            piece,
        };
    }

    const newPosL = { row: position.row, col: position.col - 1 };
    if (!hasCollision(state, newPosL, piece)) {
        return {
            state,
            position: newPosL,
            piece,
        };
    }

    const newPosR = { row: position.row, col: position.col + 1 };
    if (!hasCollision(state, newPosR, piece)) {
        return {
            state,
            position: newPosR,
            piece,
        };
    }

    return {
        state,
        position,
        piece: p,
    }
};

export const moveLeft = (state: State, position: PositionGrid, piece: Piece) => {
    // calc position
    // check for collision
    // return result if no collision
    // else return same result

    const newPos = { row: position.row, col: position.col - 1 };
    return {
        state,
        position: hasCollision(state, newPos, piece) ? position : newPos,
        piece,
    }
}

export const moveRight = (state: State, position: PositionGrid, piece: Piece) => {
    // calc position
    // check for collision
    // return result if no collision
    // else return same result
    
    const newPos = { row: position.row, col: position.col + 1 };
    return {
        state,
        position: hasCollision(state, newPos, piece) ? position : newPos,
        piece,
    }
}

export const moveDown = (state: State, position: PositionGrid, piece: Piece) => {
    // calc position
    // check for collision
    // return result if no collision
    // else return merged state + piece and generate a random piece and reset position
    
    const newPos = { row: position.row + 1, col: position.col };
    const collision = hasCollision(state, newPos, piece);
    let gameover = false;

    if (collision) {
        const result: any = {
            state: merge(state, position, piece).state as State,
            ...generateRandomPiece(),
            gameover, 
        }

        if (!hasCollision(result.state, result.position, result.piece)) {
            return result;
        } else {
            gameover = true;
            return {
                state,
                position,
                piece,
                gameover: true,
            }
        };
    } 
    return {
        state,
        position: newPos,
        piece,
        gameover
    }
}

const hasCollision = (state: State, position: PositionGrid, piece: Piece) => {
    // tslint:disable-next-line:no-console
    console.log(piece);
    for (let m=0; m<4; m++) {
        for (let n=0; n<4; n++) {
            // tslint:disable-next-line:no-console
            console.log(m + ", " + n + " = " + piece[m][n]);
            if (piece[m][n] && (
                    n+position.col < 0 || 
                    n+position.col >= state[m].length ||
                    m+position.row >= state.length ||
                    state[m+position.row][n+position.col])) {
                return true;
            }
        }
    }
    return false;
}

export const generateRandomPiece = () => {
    const mapOfPieces = {
        0: pieces.i,
        1: pieces.j,
        2: pieces.l,
        3: pieces.o,
        4: pieces.s,
        5: pieces.t,
        6: pieces.z,
    };
    
    return {
        position: {
            row: 0,
            col: 3,
        },
        piece: mapOfPieces[Math.round(Math.random() * 6)],
    }
}

export const merge = (state: State, position: PositionGrid, p: Piece) => {
    const clonedState = state.map((row) => [...row]);
    for (let m=0; m<4; m++) {
        for (let n=0; n<4; n++) {
            if (p[m][n]) {
                // tslint:disable-next-line:no-console
                console.log("ok");
                clonedState[m+position.row][n+position.col] = p[m][n];
            }
        }
    }
    
    // tslint:disable-next-line:no-console
    console.log("****");
    // tslint:disable-next-line:no-console
    console.log(clonedState);
    return {
        state: clonedState,
    }
}
