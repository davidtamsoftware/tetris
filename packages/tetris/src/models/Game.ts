import { Piece } from "./Piece";
import { PiecePosition } from "./PiecePosition";
import { Playfield } from "./Playfield";
import { Scoreboard } from "./Scoreboard";
import { Stats } from "./Stats";

export enum GameState {
    Paused,
    Active,
    GameOver,
};

export interface Game {
    playfield: Playfield;
    position: PiecePosition;
    piece: Piece;
    gameState: GameState;
    stats: Stats;
    scoreboard: Scoreboard;
    nextPiece: Piece;
}
