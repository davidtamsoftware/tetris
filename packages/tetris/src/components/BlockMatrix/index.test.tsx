import React from "react";
import renderer from "react-test-renderer";
import { l } from "tetris-core/lib/models";
import { BlockMatrix } from ".";

describe("<BlockMatrix />", () => {
    it("renders correctly", () => {
        const tree = renderer
            .create(<BlockMatrix matrix={l} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
