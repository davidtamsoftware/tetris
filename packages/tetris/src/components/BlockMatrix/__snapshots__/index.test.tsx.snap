// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BlockMatrix /> renders correctly 1`] = `
<table>
  <tbody>
    <tr>
      <td />
      <td />
      <td />
    </tr>
    <tr>
      <td>
        <div
          className="block green large"
        />
      </td>
      <td>
        <div
          className="block green large"
        />
      </td>
      <td>
        <div
          className="block green large"
        />
      </td>
    </tr>
    <tr>
      <td>
        <div
          className="block green large"
        />
      </td>
      <td />
      <td />
    </tr>
  </tbody>
</table>
`;
